### DISCUSSION TITLE
Cypress component test setup

### DISCUSSION THREAD

**Discussion by xtoolkit** [nuxt/nuxt#19304] [Sharing]
Hello
I search to setup cypress component test for nuxt version 3. But not found any result with successfully setup (like this https://github.com/cypress-io/cypress/issues/23619#issuecomment-1420459541). So I tried to solve the problems for my project and I will share the result with you.

First step change project package type to `module`:

```jsonc
// package.json
{
  "name": "my-nuxt-project",
  // ...
  "type": "module"
  // ...
}
```

Next setup nuxt vite config in cypress

```ts
// cypress.config.ts
import {defineConfig} from 'cypress';
import {loadNuxt, buildNuxt} from '@nuxt/kit';
import type {InlineConfig} from 'vite';

async function getNuxtViteConfig() {
  const nuxt = await loadNuxt({
    cwd: process.cwd(),
    dev: false,
    overrides: {
      ssr: false
    }
  });
  return new Promise<InlineConfig>((resolve, reject) => {
    nuxt.hook('vite:extendConfig', config => {
      resolve(config);
      throw new Error('_stop_');
    });
    buildNuxt(nuxt).catch(err => {
      if (!err.toString().includes('_stop_')) {
        reject(err);
      }
    });
  }).finally(() => nuxt.close());
}

export default defineConfig({
  //...
  component: {
    devServer: {
      framework: 'vue',
      bundler: 'vite',
      async viteConfig() {
        const config = await getNuxtViteConfig();

        config.plugins = config?.plugins?.filter(
          // @ts-ignore
          item => !['replace', 'vite-plugin-eslint'].includes(item.name)
        );

        if (config.server) {
          config.server.middlewareMode = false;
        }

        return config;
      }
    }
  }
});
```

And add stub Nuxt ctx and `Suspense` component in commands for support top level async support.

```ts
// cypress/support/component.ts
/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable @typescript-eslint/no-namespace */
import './commands';
import {mount} from 'cypress/vue';
import {h, Suspense} from 'vue';
import {getContext} from 'unctx';
import {
  createRouter,
  createWebHistory,
  RouterLink,
  routerKey,
  routeLocationKey
} from 'vue-router';

type MountParams = Parameters<typeof mount<any>>;
type OptionsParam = MountParams[1];
type Router = ReturnType<typeof createRouter>;

declare global {
  namespace Cypress {
    interface Chainable {
      mount(component: any, options?: OptionsParam): Chainable<any>;
      stubNuxtInject(key: string, value: unknown): void;
      stubNuxtRouter(callback: ($rotuer: Router) => void): void;
      resetNuxt(): void;
    }
  }
}

const nuxtAppCtx = getContext('nuxt-app');
const $router = createRouter({
  routes: [],
  history: createWebHistory()
});
const generateNuxtCTX = (): Record<string, any> => ({
  static: {data: {}},
  payload: {data: {}, _errors: {}},
  hook: () => () => ({}),
  hooks: {
    callHook: () => Promise.resolve()
  },
  _asyncData: {},
  _asyncDataPromises: {},
  _useHead: () => ({}),
  $router,
  _route: $router.currentRoute
});
let nuxtCTX = generateNuxtCTX();

nuxtAppCtx.set(nuxtCTX);

Cypress.Commands.add('mount', (component, options = {}) => {
  options.global = options.global || {};
  options.global.components = {RouterLink};
  options.global.provide = {
    [routerKey as symbol]: $router,
    [routeLocationKey as symbol]: $router.currentRoute.value
  };

  return mount(
    () => h(Suspense, h(component, options?.props, options?.slots)),
    options
  );
});

Cypress.Commands.add('stubNuxtInject', (key, value) => {
  nuxtCTX['$' + key] = value;
});

Cypress.Commands.add('stubNuxtRouter', callback => {
  callback($router);
});

Cypress.Commands.add('resetNuxt', () => {
  nuxtCTX = generateNuxtCTX();
  nuxtAppCtx.unset();
  nuxtAppCtx.set(nuxtCTX);

  $router.push('/');
});
```

End. now you can create cypress component test like this:
```ts
import {FragmentCustomPagePolicy} from '../../helper';

describe('<FragmentCustomPagePolicy />', () => {
  beforeEach(() => {
    cy.resetNuxt();
  });

  it('Given `CustomPage` When component `mount` Then `body` seen', () => {
    // Arrange
    cy.stubNuxtInject('policyApi', () => Promise.resolve({body: 'body'}));

    // Act
    cy.mount(FragmentCustomPagePolicy);

    // Assert
    cy.get('[data-v-app=""] > div').should('have.text', 'body');
  });
});
```

**Comment by lmiller1990** [comment]
Wow, nice job! I'm looking forward to trying this.

**Comment by JasonLandbridge** [comment]
Thank you very much!

I have it almost working but the first step:

>  First step change project package type to module: "type": "module"

Causes errors down the line (I think) when you have an import in a Vue component file:

![image](https://user-images.githubusercontent.com/15127381/228016700-1cc14e9f-60d0-415f-9acf-94d373893e56.png)

![image](https://user-images.githubusercontent.com/15127381/228016973-4ea663f6-305e-4648-8b85-9d859f56637d.png)

So in this case `import Log from 'consola';` causes the error, but if I remove that then it will complain about the next import

I found this [answer](https://stackoverflow.com/a/71022878/8205497) which mentions the `"type": "module"` being the problem, but when I remove that, then the config fails to build with `TypeError [ERR_UNKNOWN_FILE_EXTENSION]: Unknown file extension ".ts"` error. Similair to [here](https://github.com/cypress-io/cypress-docker-images/issues/673)

Any suggestions?

**Reply by xtoolkit** [reply]
Hello

Creating `nuxtAppCtx` manually means we don't see a lot of details. For example, in one of the tests, I realized that I also need to change $router.

The main question is how much we need nuxt for our component tests. In my opinion, the less it is, the faster the tests will be performed.

So, the development of nuxtAppCtx depends on the tests you write.  I changed the file `cypress/support/component.ts` in this discussions.

Please also tell me what version of nuxt you are using for this error and click on print to console and share its details so that we can solve your problem.

**Reply by JasonLandbridge** [reply]
> The main question is how much we need nuxt for our component tests. In my opinion, the less it is, the faster the tests will be performed.

Yeah I agree, the less the better. But this becomes a problem when you try to test components that use composables, or other Nuxt related functionality. For dumb components it works, but you will need Nuxt once they become more complicated. And ideally, you would like to test in an identical environment to what you run in production so I'm afraid you can't really avoid Nuxt

> So, the development of nuxtAppCtx depends on the tests you write. I changed the file `cypress/support/component.ts` in this discussions.
 
Thanks, is the project your working on open-source so I can take a better look on how it's configured?

> Please also tell me what version of nuxt you are using for this error and click on print to console and share its details so that we can solve your problem.

Thanks for the help:

```
// nuxi info
------------------------------
- Operating System: `Linux`
- Node Version:     `v18.11.0`
- Nuxt Version:     `3.3.2`
- Nitro Version:    `2.3.2`
- Package Manager:  `npm@8.19.3`
- Builder:          `vite`
- User Config:      `ssr`, `srcDir`, `devServer`, `runtimeConfig`, `modules`, `quasar`, `typescript`, `macros`, `i18n`, `css`, `alias`, `components`, `vite`, `hooks`, `telemetry`
- Runtime Modules:  `nuxt-quasar-ui@1.3.0`, `@vueuse/nuxt@9.13.0`, `@nuxt/devtools@0.2.5`, `@nuxtjs/i18n@8.0.0-beta.10`, `@vue-macros/nuxt@1.2.6`
- Build Modules:    `-`
------------------------------

```

![image](https://user-images.githubusercontent.com/15127381/228169160-8abc2bb4-10ef-4cab-a2a4-ce05f85c86cc.png)

**Reply by JasonLandbridge** [reply]
Okay I think I solved it!

The answer was described here: https://github.com/nuxt/vite/issues/56#issuecomment-1227825628

TL;DR; Use the [Vite-Plugin-CommonJS](https://github.com/originjs/vite-plugins/tree/main/packages/vite-plugin-commonjs
)

So in this case it's like this:

```ts
import { defineConfig } from 'cypress';
import { defineConfig as defineVite, InlineConfig, mergeConfig } from 'vite';
import { esbuildCommonjs, viteCommonjs } from '@originjs/vite-plugin-commonjs';
import { buildNuxt, loadNuxt } from '@nuxt/kit';

export default defineConfig({
  component: {
    viewportHeight: 1080,
    viewportWidth: 1920,
    devServer: {
      framework: 'vue',
      bundler: 'vite',
      viteConfig: async () => {
        // Source: https://github.com/nuxt/nuxt/discussions/19304
        const config = await mergeConfig(viteConfig, await nuxtViteConfig());

        // @ts-ignore
        config.plugins = config.plugins?.filter((item) => !['replace', 'vite-plugin-eslint'].includes(item.name));

        // @ts-ignore
        config.server.middlewareMode = false;

        // eslint-disable-next-line no-console
        console.log('log:entering viteconfig', config);
        return config;
      },
    },
  },
});

const viteConfig = defineVite({
  optimizeDeps: {
    disabled: true,
    esbuildOptions: { // THE FIX
      plugins: [esbuildCommonjs(['consola', 'date-fns'])],
    },
  },
 // THE FIX
  plugins: [viteCommonjs()],
});

// https://github.com/nuxt/framework/issues/6496
async function nuxtViteConfig() {
  const nuxt = await loadNuxt({
    dev: false,
    overrides: {
      srcDir: 'src',
      ssr: false,
    },
  });
  return new Promise<InlineConfig>((resolve, reject) => {
    nuxt.hook('vite:extendConfig', (config) => {
      resolve(config);
      throw new Error('_stop_');
    });
    buildNuxt(nuxt).catch((err) => {
      if (!err.toString().includes('_stop_')) {
        reject(err);
      }
    });
  }).finally(() => nuxt.close());
}
```

At this point I have it pretty much working for 80% for my components so I'm happy. For maybe others who are looking for an semi-working example, the project I'm working on is called [PlexRipper/PlexRipper](https://github.com/PlexRipper/PlexRipper/tree/NuxtV3/src/WebAPI/ClientAppV2). And is based on Nuxt v3, Quasar v2 and Vite

**Reply by xtoolkit** [reply]
Great!
Thank you for sharing solution ❤️

**Comment by lmiller1990** [comment]
You probably don't want to change to `type: module`, it will have a massive impact on a lot of your app. 

What was the error you got before using `type: module`?

**Reply by JasonLandbridge** [reply]
I got: 

```
cypress
ClientAppV2 (NuxtV3)Component Testing
[v12.8.1](https://on.cypress.io/changelog#12-8-1)

Your configFile is invalid: /mnt/PROJECTS/PlexRipper/src/WebAPI/ClientAppV2/cypress.config.ts

It threw an error when required, check the stack trace below:

[Stack trace](http://localhost:39987/__launchpad/index.html#)
TypeError [ERR_UNKNOWN_FILE_EXTENSION]: Unknown file extension ".ts" for /mnt/PROJECTS/PlexRipper/src/WebAPI/ClientAppV2/cypress.config.ts
    at new NodeError (node:internal/errors:393:5)
    at Object.getFileProtocolModuleFormat [as file:] (node:internal/modules/esm/get_format:79:11)
    at defaultGetFormat (node:internal/modules/esm/get_format:121:38)
    at defaultLoad (node:internal/modules/esm/load:81:20)
    at nextLoad (node:internal/modules/esm/loader:163:28)
    at ESMLoader.load (node:internal/modules/esm/loader:605:26)
    at ESMLoader.moduleProvider (node:internal/modules/esm/loader:457:22)
    at new ModuleJob (node:internal/modules/esm/module_job:63:26)
    at ESMLoader.#createModuleJob (node:internal/modules/esm/loader:480:17)
    at ESMLoader.getModuleJob (node:internal/modules/esm/loader:434:34)
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async Promise.all (index 0)
    at async ESMLoader.import (node:internal/modules/esm/loader:530:24)
    at async importModuleDynamicallyWrapper (node:internal/vm/module:438:15)
    at async loadFile (/home/jason/.cache/Cypress/12.8.1/Cypress/resources/app/node_modules/@packages/server/lib/plugins/child/run_require_async_child.js:106:14)
    at async EventEmitter. (/home/jason/.cache/Cypress/12.8.1/Cypress/resources/app/node_modules/@packages/server/lib/plugins/child/run_require_async_child.js:116:32)
```

**Reply by JasonLandbridge** [reply]
I ended up just accepting the `type: module` property in the `package.json` which fixed the issue. I read up on the meaning behind it, as I initially agreed that it might have a big impact. Based on this [answer on stack overflow](https://stackoverflow.com/a/62554884/8205497), it seems that with the higher node versions (16+) it's not a problem to do, and with Nuxt 3 I'm on Node 18 anyway. I also only use `import` instead of `require`, so this seems acceptable to me. 

Please let me know if I misunderstood anything and thanks!

**Reply by lmiller1990** [reply]
Glad to here you got it working. If you look in the Cypress repo for your error, you can find few common fixes for it (that are not adding type=module). You might need to use `.js` or change your `tsconfig.json` to specify `module: 'commonjs'`.

If you have a repo you can't get working, you can share a minimal reproduction and I can look. It sounds like you've found something that works for you, though.

**Reply by Floppy012** [reply]
Sadly I cannot use the `type: module` way in my package.json as it causes other tools (such as `lint-staged`) to fail. If I don't have the `type: module` set I get the following error:

```
Error [ERR_PACKAGE_PATH_NOT_EXPORTED]: No "exports" main defined in /project/node_modules/@nuxt/kit/package.json
    at new NodeError (node:internal/errors:399:5)
    at exportsNotFound (node:internal/modules/esm/resolve:361:10)
    at packageExportsResolve (node:internal/modules/esm/resolve:641:13)
    at resolveExports (node:internal/modules/cjs/loader:565:36)
    at Function.Module._findPath (node:internal/modules/cjs/loader:634:31)
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:1061:27)
    at Function.Module._resolveFilename (/root/.cache/Cypress/12.9.0/Cypress/resources/app/node_modules/tsconfig-paths/lib/register.js:75:40)
    at Function.Module._resolveFilename.sharedData.moduleResolveFilenameHook.installedValue [as _resolveFilename] (/root/.cache/Cypress/12.9.0/Cypress/resources/app/node_modules/@cspotcode/source-map-support/source-map-support.js:811:30)
    at Function.Module._load (node:internal/modules/cjs/loader:920:27)
    at Module.require (node:internal/modules/cjs/loader:1141:19)
    at require (node:internal/modules/cjs/helpers:110:18)
    at Object. (/project/cypress.config.ts:2:1)
    at Module._compile (node:internal/modules/cjs/loader:1254:14)
    at Module.m._compile (/root/.cache/Cypress/12.9.0/Cypress/resources/app/node_modules/ts-node/dist/index.js:857:29)
    at Module._extensions..js (node:internal/modules/cjs/loader:1308:10)
    at Object.require.extensions. [as .ts] (/root/.cache/Cypress/12.9.0/Cypress/resources/app/node_modules/ts-node/dist/index.js:859:16)
    at Module.load (node:internal/modules/cjs/loader:1117:32)
    at Function.Module._load (node:internal/modules/cjs/loader:958:12)
    at Module.require (node:internal/modules/cjs/loader:1141:19)
    at require (node:internal/modules/cjs/helpers:110:18)
    at loadFile (/root/.cache/Cypress/12.9.0/Cypress/resources/app/node_modules/@packages/server/lib/plugins/child/run_require_async_child.js:89:14)
    at EventEmitter. (/root/.cache/Cypress/12.9.0/Cypress/resources/app/node_modules/@packages/server/lib/plugins/child/run_require_async_child.js:116:38)
```

I have tried:

1. Using `cypress.config.js`, `cypress.config.cjs`,  `cypress.config.mjs` (the latter brought me one step further but then failed with undefined token `export`)
2. Specifying `module: 'commonjs'` in `tsconfig.json`

The only way I got it working was setting `type: module` in package.json but that broke everything else.

**Reply by lmiller1990** [reply]
`type: module` should work fine. Can you share your `cypress.config.js`? What error do you encounter?

**Reply by Floppy012** [reply]
> `type: module` should work fine. Can you share your `cypress.config.js`? What error do you encounter?

**Error**
```
$ /path/to/project/node_modules/.bin/lint-staged
ReferenceError: require is not defined in ES module scope, you can use import instead
This file is being treated as an ES module because it has a '.js' file extension and '/path/to/project/package.json' contains "type": "module". To treat it as a CommonJS script, rename it to use the '.cjs' file extension.
    at file:///path/to/project/.lintstagedrc.js:1:14
    at ModuleJob.run (node:internal/modules/esm/module_job:193:25)
    at async Promise.all (index 0)
    at async ESMLoader.import (node:internal/modules/esm/loader:533:24)
    at async Object.load (/path/to/project/node_modules/lint-staged/node_modules/lilconfig/dist/index.js:167:19)
    at async loadConfig (file:///path/to/project/node_modules/lint-staged/lib/loadConfig.js:65:20)
    at async Promise.all (index 0)
    at async searchConfigs (file:///path/to/project/node_modules/lint-staged/lib/searchConfigs.js:84:3)
    at async runAll (file:///path/to/project/node_modules/lint-staged/lib/runAll.js:129:24)
    at async lintStaged (file:///path/to/project/node_modules/lint-staged/lib/index.js:108:17)
ReferenceError: require is not defined in ES module scope, you can use import instead
This file is being treated as an ES module because it has a '.js' file extension and '/path/to/project/package.json' contains "type": "module". To treat it as a CommonJS script, rename it to use the '.cjs' file extension.
    at file:///path/to/project/.lintstagedrc.js:1:14
    at ModuleJob.run (node:internal/modules/esm/module_job:193:25)
    at async Promise.all (index 0)
    at async ESMLoader.import (node:internal/modules/esm/loader:533:24)
    at async Object.load (/path/to/project/node_modules/lint-staged/node_modules/lilconfig/dist/index.js:167:19)
    at async loadConfig (file:///path/to/project/node_modules/lint-staged/lib/loadConfig.js:65:20)
    at async Promise.all (index 0)
    at async searchConfigs (file:///path/to/project/node_modules/lint-staged/lib/searchConfigs.js:84:3)
    at async runAll (file:///path/to/project/node_modules/lint-staged/lib/runAll.js:129:24)
    at async lintStaged (file:///path/to/project/node_modules/lint-staged/lib/index.js:108:17)
✖ No valid configuration found.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
husky - pre-commit hook exited with code 1 (error)
```

My last post was on a Friday and I haven't really had time to look into it. But now reading the error it would work if I changed the ending of any `.js` config file to `.cjs`. But that feels pretty ugly :/

**cypress.config.ts**

<details>
<summary>Click me</summary>

```typescript
import {defineConfig} from "cypress";
import {loadNuxt, buildNuxt} from '@nuxt/kit';
import type {InlineConfig} from 'vite';

/**
 * @see https://github.com/nuxt/nuxt/discussions/19304
 */
async function getNuxtViteConfig() {
    const nuxt = await loadNuxt({
        cwd: process.cwd(),
        dev: false,
        overrides: {
            ssr: false
        }
    });
    return new Promise<InlineConfig>((resolve, reject) => {
        nuxt.hook('vite:extendConfig', config => {
            resolve(config);
            throw new Error('_stop_');
        });
        buildNuxt(nuxt).catch(err => {
            if (!err.toString().includes('_stop_')) {
                reject(err);
            }
        });
    }).finally(() => nuxt.close());
}

export default defineConfig({
    downloadsFolder: ".cypress/downloads",
    screenshotsFolder: ".cypress/screenshots",
    videosFolder: ".cypress/videos",
    supportFolder: "tests/Cypress/Support",
    viewportWidth: 1920,
    viewportHeight: 1080,

    e2e: {
        supportFile: "tests/Cypress/Support/e2e.ts",
        specPattern: "tests/Cypress/E2E/Tests/**/*.cy.{js,jsx,ts,tsx}",
        fixturesFolder: "tests/Cypress/E2E/Fixtures",
        experimentalStudio: true,
        setupNodeEvents(on, config) {
            // implement node event listeners here
        },
    },

    component: {
        indexHtmlFile: 'tests/Cypress/Support/component-index.html',
        supportFile: "tests/Cypress/Support/component.ts",
        fixturesFolder: 'tests/Cypress/Component/Fixtures',
        specPattern: "tests/Cypress/Component/Tests/**/*.cy.{js,jsx,ts,tsx}",
        devServer: {
            framework: "vue",
            bundler: "vite",
            /**
             * @see https://github.com/nuxt/nuxt/discussions/19304
             */
            async viteConfig() {
                const config = await getNuxtViteConfig();

                config.plugins = config?.plugins?.filter(
                    // @ts-ignore
                    item => !['replace', 'vite-plugin-eslint'].includes(item.name)
                );

                if (config.server) {
                    config.server.middlewareMode = false;
                }

                return config;
            }
        },
    },
});
```

</details>

**Reply by lmiller1990** [reply]
I am a bit confused. Your repro:

```
/path/to/project/node_modules/.bin/lint-staged
...
This file is being treated as an ES module because it has a '.js' file extension and '/path/to/project/package.json' contains "type": "module". To treat it as a CommonJS script, rename it to use the '.cjs' file extension.
    at file:///path/to/project/.lintstagedrc.js:1:14
...
```

Why is lint-staged running? Are you encountering this error when you run `npx cypress open`?

**Reply by Floppy012** [reply]
Oh. I'm sorry for the confusion. Cypress runs fine with the `type: module` thing. But all the other tools (lint-staged, eslint, prettier, etc.) stop working with these errors.

**Reply by lmiller1990** [reply]
That's not ideal... but those tools should work fine with `type: module`. Cypress should work with both, too.

If you can provide a minimal repo with an error (either with or without type: module), I can take a look.

**Reply by Floppy012** [reply]
Thanks. Here you go: https://github.com/floppy012-repros/nuxt-cypress

**Reply by lmiller1990** [reply]
Hey, I am sorry I didn't get back to you. I'm not sure I have time to do this right now -- I'll try to loop back to it in the future.

**Reply by Floppy012** [reply]
No worries. I used the `cjs` workaround i posted before and its working. So just a minor inconvenience.

**Comment by vedmant** [comment]
Tried this config but having an error in Cypress:

```
[plugin:vite:import-analysis] Failed to parse source for import analysis because the content contains invalid JS syntax. Install @vitejs/plugin-vue to handle .vue files.
```

**Comment by abdul-alhasany** [comment]
@vedmant You need to install and add `@vitejs/plugin-vue`.

```diff
// cypress.config.ts
import {defineConfig} from 'cypress';
import {loadNuxt, buildNuxt} from '@nuxt/kit';
import type {InlineConfig} from 'vite';
+ import vuePlugin from '@vitejs/plugin-vue';

// ...
export default defineConfig({
  //...
  component: {
    devServer: {
      framework: 'vue',
      bundler: 'vite',
      async viteConfig() {
        const config = await getNuxtViteConfig();

+       config.plugins.push(vuePlugin()); // Add it here

        config.plugins = config?.plugins?.filter(
          // @ts-ignore
          item => !['replace', 'vite-plugin-eslint'].includes(item.name)
        );

        if (config.server) {
          config.server.middlewareMode = false;
        }

        return config;
      }
    }
  }
});
```

**Comment by halkony** [comment]
Unfortunately, none of the solutions in this thread seem to work with my component libraries, even if the UI element is explicitly imported inside of the SFC.
